bounded_sum_known_n/BoundedSumKnownN.tex

```diff
@@ -0,0 +1,394 @@
+    \item The Rust code does not check for possible overflow errors when performing the arithmetic operation $\texttt{U}-\texttt{L}$, whereas the current pseudocode includes a \texttt{checked\_sub} for it.
+    \item The Rust code does not include an \texttt{inf\_cast} in the stability relation, which is necessary for type correctness, but there is already a PR that resolves this discrepancy: \url{https://github.com/opendp/opendp/blob/493fe87273adca3294551faa927039b755c1756e/rust/opendp/src/trans/sum/mod.rs#L65-L94}.
+    \item The previous PR also includes the new user-specified type trait \texttt{InfDiv}, which the current pseudocode does not include.
+\end{itemize}
+
+These changes will soon be applied to the Rust code. The pseudocode also uses the variable \texttt{n} instead of Rust's \texttt{size}, which is more conventional for proof writing.
+
+
+\section{Algorithm Implementation}
+\subsection{Code in Rust}
+The current OpenDP library contains the transformation \texttt{make\_bounded\_sum\_n} implementing the bounded sum function with known $n$. This is defined in lines 29-49 of the file \texttt{mod.rs} in the Git repository\footnote{Updated on September 20, 2021.} (\url{https://github.com/opendp/opendp/blob/8bbb0fab1da9b86c50235a36d7026189be43b1ab/rust/opendp/src/trans/sum/mod.rs#L29-L49}).
+
+%\begin{figure}[ht]
+    %\includegraphics[width=15cm]{bounded_sum_2.png}
+    %\centering
+    %\label{fig:code}
+%\end{figure}
+
+\begin{figure}[ht]
+    \includegraphics[width=15cm]{bounded_sum_n_new.png}
+    \centering
+    \label{fig:code}
+\end{figure}
+
+\subsection{Pseudocode in Python}
+\label{sec:pseudocode}
+We present a simplified Python-like pseudocode of the Rust implementation below. The necessary definitions for the pseudocode can be found in Section \ref{sec:versioned-docs}.
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Variable \texttt{n} must be of type \texttt{usize}.
+        \item Type \texttt{T} must have traits \texttt{DistanceConstant(IntDistance)}, \texttt{TotalOrd}, \texttt{CheckedMul}, \texttt{CheckNull}, \texttt{Sum(Output=T)}, \texttt{Sub(Output=T)},  and \texttt{ExactIntCast(usize)}.
+        \item \texttt{IntDistance} must have trait \texttt{InfCast(T)}. (Note that this bullet point is not needed in this proof, but it is needed in the code so a hint can be constructed; otherwise a binary search would be needed to construct the hint.)
+        \item Variables \texttt{U} and \texttt{L} must be of type \texttt{T}, and we must have $\texttt{L} \leq \texttt{U}$.
```

Comment:
I think L<=U is fine as a precondition (don't need it to be L<U) - this just means that the interval domain has only one element, unless @Shoeboxam feels differently.

Reply:
Based on the comment, it looks like no changes are needed. Please let me know if I misinterpreted the feedback.

Also, we looked pretty closely at the code yesterday, and it looks like L<=U (rather than L<U) is correct.

Reply:
right, I don't see any needed changes.  resolving the conversation.

Reply:
Yes, Mike confirmed L<=U